src/dataMap/metaManager/lazyFactoryMap.js

```diff
@@ -0,0 +1,265 @@
+  constructor(valueFactory) {
+    this.valueFactory = valueFactory;
+    /**
+     * An array which contains data.
+     *
+     * @type {Array}
+     */
+    this.data = [];
+    /**
+     * An array of indexes where the key of the array is mapped to the value which points to the
+     * specific position of the data array.
+     *
+     * @type {Number[]}
+     */
+    this.index = [];
+    /**
+     * An array of indexes where points to the data items which can be replaced by obtaining new
+     * ones. The "holes" are a side effect of deleting entries.
+     *
+     * @type {Number[]}
+     */
+    this.holes = [];
+  }
+
+  /**
+   * Gets or if data not exist creates and returns new data.
+   *
+   * @param {Number} key The item key as zero-based index.
+   * @returns {*}
+   */
+  obtain(key) {
+    assert(() => isFiniteUnsignedNumber(key), 'Expecting an unsigned finite number.');
+
+    const dataIndex = this._getStorageIndexByKey(key);
+    let result;
+
+    if (dataIndex >= 0) {
+      result = this.data[dataIndex];
+
+      if (result === void 0) {
+        result = this.valueFactory(key);
+        this.data[dataIndex] = result;
+      }
+    } else {
+      result = this.valueFactory(key);
+
+      if (this.holes.length > 0) {
+        const reuseIndex = this.holes.pop();
+
+        this.data[reuseIndex] = result;
+        this.index[key] = reuseIndex;
+      } else {
+        this.data.push(result);
+        this.index[key] = this.data.length - 1;
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * Inserts an empty data to the map. This method creates an empty space for obtaining
+   * new data.
+   *
+   * @param {Number} key The key as volatile zero-based index at which to begin inserting space for new data.
+   * @param {Number} [amount=1] Ammount data to insert.
+   */
+  insert(key, amount = 1) {
+    assert(() => (isFiniteUnsignedNumber(key) || isNullish(key)), 'Expecting an unsigned finite number or null/undefined argument.');
+
+    const newIndexes = [];
+    const dataLength = this.data.length;
+
+    for (let i = 0; i < amount; i++) {
+      newIndexes.push(dataLength + i);
+      this.data.push(void 0);
+    }
+
+    this.index.splice(isNullish(key) ? this.index.length : key, 0, ...newIndexes);
+  }
+
+  /**
+   * Removes (soft remove) data from "index" and according to the amount of data.
+   *
+   * @param {Number} key The key as volatile zero-based index at which to begin removing the data.
+   * @param {Number} [amount=1] Ammount data to remove.
+   */
+  remove(key, amount = 1) {
+    assert(() => (isFiniteUnsignedNumber(key) || isNullish(key)), 'Expecting an unsigned finite number or null/undefined argument.');
+
+    const removed = this.index.splice(isNullish(key) ? this.index.length - amount : key, amount);
+
+    for (let i = 0; i < removed.length; i++) {
+      const removedIndex = removed[i];
+
+      if (typeof removedIndex === 'number') {
+        this.holes.push(removedIndex);
+      }
```

Comment:
Is it really needed? I’m not sure if there are some cases when code inside the `if` won’t be executed. Few lines above there is some ternary operator which set some number to the `removed` variable?

Reply:
The `this.index` array currently keeps numeric values with undefined values. It has to be checked if the removed value are indeed number value.